Revised config and code quality improvements

clessig opened the pull request:
## Description

Revise config to nested dict; simplify code where possible and where changes are necessary anyway.

This PR also enables a more flexible combination of different loss terms, e.g. of a physical space and latent loss, as demonstrated in the default config. It also decouples training and validation and test as much as possible, so that one can have different objectives for these.

## Issue Number

Closes #1534 
Closes #1535

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

clessig committed:
Partially revised config; model is still missing but proper setup of training_config and validation_config

clessig committed:
Changes necessary due to changed position of time keys and of run_id

clessig committed:
Handling of multiple loss terms / target_aux_calculators and non-LossPhysical ones.

clessig committed:
Changed position of run_id in config

clessig committed:
Add function to extract batch size from mode_cfg

clessig committed:
Changed position of run_id in config

clessig committed:
Changes due to revised config. Also proper handling of target_aux_calculator and various other details cleaned up

clessig committed:
Revised config structure, in particular for losses, and related changes

clessig committed:
Add missing copyright and minor changed to to_device()

clessig committed:
Moved sanity checking from trainer here. Also learning_rate sub_part of config is passed to LRScheduler, which leads to major simplifications

clessig committed:
Minor cleanups

clessig committed:
Changes due to changed structure of losses in config

clessig committed:
Changes due to changed structure of losses in config

clessig committed:
Minor changes due to changed position of run_id in config

clessig committed:
Minor changes to accomodate new config, in particular target_aux_calculator config

clessig committed:
Support batch_size > 1. Clean up of various smaller parts

clessig committed:
Clean up and implementation for batch_size > 1.

clessig committed:
Fix to sharding problem with FSDP2

clessig committed:
Removed scatter offset computation which now happens on the fly in the model

clessig committed:
Changes for revised config, simplify overall where possible

clessig committed:
Fix issues with source-target sample generation and matching. Work in progress

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Type hint

clessig committed:
Linting

clessig commented:
forecast, ERA5 + NPP-ATMS + SYNOP


num_samples/batch_size = 1: (520709 : g479xfk6)

0: LossPhysical.ERA5.mse.avg : 9.4735E-02       
0: LossPhysical.ERA5.mae.avg : 1.9303E-01       
0: LossPhysical.NPPATMS.mse.avg : 1.1044E-02    
0: LossPhysical.NPPATMS.mae.avg : 2.9273E-02    
0: LossPhysical.SurfaceCombined.mse.avg : 2.6745E-01    
0: LossPhysical.SurfaceCombined.mae.avg : 2.8849E-01    
0: LossPhysical.loss_avg : 1.8611E-01 



num_samples/batch_size = 2: 520711 : ypix0nr7

0: LossPhysical.ERA5.mse.avg : 1.1328E-01       
0: LossPhysical.ERA5.mae.avg : 2.1473E-01       
0: LossPhysical.NPPATMS.mse.avg : 1.8830E-02    
0: LossPhysical.NPPATMS.mae.avg : 4.5836E-02    
0: LossPhysical.SurfaceCombined.mse.avg : 2.9414E-01    
0: LossPhysical.SurfaceCombined.mae.avg : 3.0358E-01    
0: LossPhysical.loss_avg : 2.0496E-01

clessig committed:
Linting

clessig committed:
Linting

clessig committed:
Renamed loss keys for consistency

iluise committed:
implement reader merge

sophie-xhonneux commented on config/default_config.yml:
```diff
@@ -45,24 +44,19 @@ pred_adapter_kv: False
 pred_mlp_adaln: True
-num_class_tokens: 1
-num_register_tokens: 7
+num_class_tokens: 0
```
I think the default register tokens should probably not be 0

clessig replied:
I think the defaul_config should be a relatively minimal example to run the model. So I don't think we need register tokens. But we can have a discussion in #1561

sophie-xhonneux commented on config/default_config.yml:
```diff
@@ -45,24 +44,19 @@ pred_adapter_kv: False
 fe_with_qk_lnorm: True
 fe_layer_norm_after_blocks: []  # Index starts at 0. Thus, [3] adds a LayerNorm after the fourth layer
 fe_impute_latent_noise_std: 0.0  # 1e-4
+forecast_att_dense_rate: 1.0
```
I think maybe adding a comment here or in the code would be good to explain that this can cause an error when not 1.0 due to local attention not existing

sophie-xhonneux commented on src/weathergen/datasets/multi_stream_data_sampler.py:
```diff
@@ -71,45 +71,54 @@ class MultiStreamDataSampler(torch.utils.data.IterableDataset):
-        end_date: np.datetime64,
-        batch_size: int,
-        samples_per_mini_epoch: int,
+        mode_cfg: dict,
```
I personally find this name unintuitive. Maybe we can at least a comment explain why we need cf and mode_cfg and what each contains?

clessig replied:
I agree but I couldn't think of a term that captures training/test/validation

sophie-xhonneux commented on src/weathergen/model/engines.py:
```diff
@@ -91,32 +92,40 @@ def forward(self, batch, pe_embed):
+
+        # switch from stream to cell-based ordering and apply per cell positional encoding
+
+        # computer scatter index across batch items and input steps
```
I might be out of the loop, how come we changed the ordering from stream to cell based ordering?

clessig replied:
This always happened here. Embedding is per stream since it happens independent for each cell and stream, subsequent computations combine tokens from different streams per cell.

sophie-xhonneux commented on src/weathergen/model/engines.py:
```diff
@@ -395,7 +404,7 @@ def __init__(self, cf: Config, num_healpix_cells: int, dim_aux: int = None) -> N
 
         global_rate = int(1 / self.cf.forecast_att_dense_rate)
-        if self.cf.forecast_policy is not None:
+        if mode_cfg.get("forecast", {}).get("policy") is not None:
```
should we have a default for `.get("policy")`

clessig replied:
Default is `null` in yaml, i.e. is None. In this case forecasting is disabled. I think that's sensible.

sophie-xhonneux commented on src/weathergen/model/model_interface.py:
```diff
@@ -256,20 +257,20 @@ def get_model(cf: Config, training_mode: TrainingMode, dataset, overrides):
 
-def get_target_aux_calculator(cf: Config, dataset, model, device, **kwargs):
+def get_target_aux_calculator(
+    cf: Config, cfg: OmegaConf, dataset, model, device, batch_size_per_gpu, **kwargs
```
the `cf` and `cfg` is confusing, we should think if we have a better alternative

clessig replied:
Yes, I changed the name to `loss_term_cfg`.

sophie-xhonneux commented on src/weathergen/model/model_interface.py:
```diff
@@ -279,9 +280,9 @@ def get_target_aux_calculator(cf: Config, dataset, model, device, **kwargs):
-        target_aux = EMATeacher(
-            model, ema_model, get_batch_size(cf, cf.world_size_original), **cf.training_config
-        )
+        batch_size = cf.get("world_size_original", cf.get("world_size")) * batch_size_per_gpu
```
shouldn't we use `get_batch_size_from_config` here like in the trainer?

clessig replied:
No, because we can have a different batch size for training and validation. We pass the correct one as an argument.

clessig committed:
Long list of fixes and improvements

clessig committed:
Enabled support for minimal configs without rate

clessig commented:
Minimal config in `training_config`:

```
  losses : {
    "physical": {
        type: LossPhysical,
        loss_fcts: { "mse": { }, },
        },
    }
  
  model_input: {
    "forecasting" : {
      masking_strategy: "forecast",
    }
  }
```

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Fixed validation. validation_io still broken

clessig committed:
Fixed linting

clessig committed:
Fixed problem with target filtering for loss computation for SSL losses

iluise committed:
working version of merge reader

iluise committed:
linter

iluise committed:
lint

iluise committed:
fix lead time

iluise committed:
Merge branch 'develop' into iluise/fix-lead-time

clessig committed:
Re-instantiated per loss-fct source/target correspondences. Introduced idx and correspondence fields to per sample meta-data which makes correct correspondence for loss computation much easier.

clessig committed:
Fixed problem with undefined variable

clessig committed:
Revised config

clessig committed:
Fixed bug with forecasting

clessig committed:
Added sanity check for config

clessig committed:
Fix bug with duplicate targets

clessig committed:
Linting

clessig committed:
Fixed problem when losses is not specified in validation config

sophie-xhonneux committed:
Fix DINOv2

clessig committed:
Removed temporary patches; fixed properly in 10b7a28

clessig committed:
Linting

clessig committed:
Patched validation IO. Needs to be fixed properly.

clessig committed:
Removed unused function

clessig committed:
Improved variable naming

clessig committed:
Improved encapsulation of functionality: total_batch_size

clessig committed:
Fixed broken inference

clessig committed:
Fixed problem with test where incorrect config was used

clessig committed:
Fixed processing and handling of spoof flag in loss calculation

clessig committed:
Fixed problem with pure masking where forecast_steps were 0. Removed duplicate function introduced through merge problem

clessig committed:
Fixed bug when output_streams is specified explicitly

clessig committed:
Corrected config param for number of samples

clessig committed:
Fixed bug in handling of spoof weight

clessig committed:
Improved clarity of logging statements

clessig committed:
Improved logging msgs

sophie-xhonneux committed:
Fix sinkhorn knopp

sophie-xhonneux committed:
Fix sinkhorn in multi-GPU mode

clessig committed:
Removed some old comments

clessig committed:
Fixed inference overwrites

clessig committed:
Merge branch 'clessig/develop/fix_config_1534' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Fixing empty output when masking

clessig committed:
Intermediate stage to re-enable integration test

clessig committed:
Adjusted thresholds

clessig committed:
Renaming

clessig committed:
Removing old config files

clessig committed:
Adding copyright

clessig committed:
Revised default_config. This is a minimal example config for simple training towards forecasting

clessig committed:
Changed multiprocessing param

clessig committed:
Adapation for new position of multiprocessing param

clessig committed:
Adding example config that combines an SSL and physical loss term

clessig committed:
More cleanup

clessig committed:
Restoring some default values

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Restoring default for decoder_type

iluise committed:
Merge branch 'develop' into iluise/fix-lead-time

iluise committed:
update to develop

iluise committed:
lint

clessig committed:
Merge branch 'iluise/fix-lead-time' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Fixed problem where parameter was expected in old config place

clessig committed:
Fixed linting

clessig committed:
Simplified interface

clessig committed:
Re-enabled forecast step and location weighting

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Merge branches 'develop' and 'clessig/develop/fix_config_1534' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Merge branch 'develop' of github.com:ecmwf/WeatherGenerator into clessig/develop/fix_config_1534

clessig committed:
Linting

clessig committed:
Using new option to have validate_before_training as an int arg that allows to specify number of samples;
Added copyright statement

clessig committed:
Added option to have validate_before_training as int argument (specifyiung the number of samples).
Fixed some minor subtle problems in validate() to fully distinguish validation and testing.

clessig committed:
Refactored correspondence parsing

sophie-xhonneux committed:
Sophiex/dev/teacher overrides (#1557)

* Add option to modify teacher TODO fix ema update

* Fix EMA under teacher and student model differences

* Attempt to revert newline

* Raise error if teacher has weights not in student

* Clessig/sophiex/dev/teacher overrides (#1585)

* Simplified error message

* Added support for target_and_aux configs

* Fix bug that validation EMA params are not used

* Removing unused/superfluous function

* Removed debug statement

* Changed config so that target_aux params are specified as dict at the appropriate place

---------

Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

clessig committed:
Fixed missing default value

sophie-xhonneux committed:
Bilinear decoder: adapt code for batchsize > 1 (#1592)

* Adapt code for batchsize > 1

* Fixed comment

---------

Co-authored-by: Christian Lessig <christian.lessig@ecmwf.int>

clessig committed:
Changed defaults

clessig committed:
Linting

clessig committed:
Fixed linting issue

clessig committed:
Reverting to ERA5-only as default

clessig committed:
Fixed problem with train_continue

clessig commented:
Default integration test is working. train_continue and inference are also working.
